src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -62,6 +66,171 @@ public static Document Current
+                this.InternalDocument.Delete(materials);
+            List<ElementId> purgedMaterials = new List<ElementId>();
+            purgedMaterials.AddRange(PurgeMaterialAssets(nonPurgedElements));
+            purgedMaterials.AddRange(materials);
+
+            return purgedMaterials;
+        }
+
+        private List<Autodesk.Revit.DB.ElementId> PurgeMaterialAssets(List<Autodesk.Revit.DB.ElementId> elementIds)
+        {
+            List<ElementId> appearanceAssets = new FilteredElementCollector(this.InternalDocument).OfClass(typeof(AppearanceAssetElement))
+                                                                                      .ToElementIds().ToList();
+
+            List<ElementId> propertySet = new FilteredElementCollector(this.InternalDocument).OfClass(typeof(PropertySetElement))
+                                                                                                      .ToElementIds().ToList();
+
+            for (int i = 0; i < elementIds.Count(); i++)
+            {
+                Autodesk.Revit.DB.Material material = this.InternalDocument.GetElement(elementIds[i]) as Autodesk.Revit.DB.Material;
+                propertySet.Remove(material.ThermalAssetId);
+                propertySet.Remove(material.StructuralAssetId);
+                appearanceAssets.Remove(material.AppearanceAssetId);
+            }
+
+            if (appearanceAssets.Count > 0)
+                this.InternalDocument.Delete(appearanceAssets);
+            if (propertySet.Count > 0)
+                this.InternalDocument.Delete(propertySet);
+
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            purgedElementIds.AddRange(appearanceAssets);
+            purgedElementIds.AddRange(propertySet);
+            return purgedElementIds;
+        }
+
+        private List<ElementId> PurgeElements(List<PerformanceAdviserRuleId> performanceAdviserRuleId, bool runRecursively)
+        {
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            List<ElementId> purgableElementIds = GetPurgeableElementIds(performanceAdviserRuleId);
+            
+            if (!runRecursively)
+            {
+                if (purgableElementIds == null)
+                    return purgedElementIds;
+                purgedElementIds.AddRange(purgableElementIds);
+                this.InternalDocument.Delete(purgableElementIds);
+                return purgedElementIds;
+            }
+            if (purgableElementIds == null)
+                return purgedElementIds;
+
+            purgedElementIds.AddRange(purgableElementIds);
+            this.InternalDocument.Delete(purgableElementIds);
+            purgedElementIds.AddRange(PurgeElements(performanceAdviserRuleId, runRecursively));
+
+            return purgedElementIds;
+        }
+
+        private List<Autodesk.Revit.DB.ElementId> GetPurgeableElementIds(List<PerformanceAdviserRuleId> performanceAdviserRuleId)
```

Comment:
this doesn't need the `performanceAdviserRuleId` as an input either, it could simply call the utility method directly.
this saves passing this list between 3 layers of methods to be used only here.